flink-libraries/flink-table/src/main/scala/org/apache/flink/table/runtime/aggregate/AggregateUtil.scala

```diff
@@ -363,199 +342,112 @@ object AggregateUtil {
+      throw new TableException(
+        "Could not find output field in input data type or aggregate functions.")
     }
+
+    val accumulatorRowType = createAccumulatorRowType(inputType, groupKeysIndex, aggregates)
+    val aggFunction = new AggregateAggFunction(
+      aggregates,
+      aggFields,
+      aggregateMapping,
+      groupKeysIndex,
+      groupKeysMapping,
+      outputType.getFieldCount)
+
+    (aggFunction, accumulatorRowType)
   }
 
   /**
-    * Create a [[WindowFunction]] to finalize incrementally pre-computed window aggregates.
+    * Return true if all aggregates can be partially merged. False otherwise.
     */
-  private[flink] def createWindowIncrementalAggregationFunction(
-      window: LogicalWindow,
-      namedAggregates: Seq[CalcitePair[AggregateCall, String]],
+  private[flink] def doAllSupportPartialMerge(
+      aggregateCalls: Seq[AggregateCall],
       inputType: RelDataType,
-      outputType: RelDataType,
-      groupings: Array[Int],
-      properties: Seq[NamedWindowProperty])
-    : WindowFunction[Row, Row, Tuple, DataStreamWindow] = {
+      groupKeysCount: Int): Boolean = {
 
-    val aggregates = transformToAggregateFunctions(
-      namedAggregates.map(_.getKey),inputType,groupings.length)._2
-
-    val (groupingOffsetMapping, aggOffsetMapping) =
-      getGroupingOffsetAndAggOffsetMapping(
-        namedAggregates,
-        inputType,
-        outputType,
-        groupings)
-
-    val finalRowArity = outputType.getFieldCount
+    val aggregateList = transformToAggregateFunctions(
+      aggregateCalls,
+      inputType,
+      groupKeysCount)._2
 
-    if (isTimeWindow(window)) {
-      val (startPos, endPos) = computeWindowStartEndPropertyPos(properties)
-      new IncrementalAggregateTimeWindowFunction(
-        aggregates,
-        groupingOffsetMapping,
-        aggOffsetMapping,
-        finalRowArity,
-        startPos,
-        endPos)
-      .asInstanceOf[WindowFunction[Row, Row, Tuple, DataStreamWindow]]
-    } else {
-      new IncrementalAggregateWindowFunction(
-        aggregates,
-        groupingOffsetMapping,
-        aggOffsetMapping,
-        finalRowArity)
-    }
+    doAllSupportPartialMerge(aggregateList)
   }
 
   /**
-    * Return true if all aggregates can be partially computed. False otherwise.
+    * Return true if all aggregates can be partially merged. False otherwise.
     */
-  private[flink] def doAllSupportPartialAggregation(
-    aggregateCalls: Seq[AggregateCall],
-    inputType: RelDataType,
-    groupKeysCount: Int): Boolean = {
-    transformToAggregateFunctions(
-      aggregateCalls,
-      inputType,
-      groupKeysCount)._2.forall(_.supportPartial)
+  private[flink] def doAllSupportPartialMerge(
+      aggregateList: Array[TableAggregateFunction[_ <: Any]]): Boolean = {
+    var ret: Boolean = true
+    var i: Int = 0
+    while (i < aggregateList.length && ret) {
+      ret = ifMethodExitInFunction("merge", aggregateList(i))
```

Comment:
it will break the loop once ret = false, so there should not be any override. This should be exactly same as forall (I was intend to ensure the performance without traveling the entire list, but it seems forall behaviors in the same way)